pkg/auth/handler/sso/sso.go

```diff
@@ -196,17 +179,42 @@ func (h respHandler) handleLogin(
 	return
 }
 
+func (h respHandler) findPrincipal(oauthAuthInfo sso.AuthInfo) (*oauth.Principal, error) {
+	// find oauth principal from principal_oauth
+	principal, err := h.OAuthAuthProvider.GetPrincipalByProviderUserID(oauthAuthInfo.ProviderName, oauthAuthInfo.ProviderUserID)
+	if err != nil {
+		if err != skydb.ErrUserNotFound {
+			return nil, err
+		}
+	} else {
+		return principal, nil
+	}
+
+	// if oauth principal doesn't exist, try to link existed password principal
+	if valid := h.PasswordAuthProvider.IsAuthDataValid(oauthAuthInfo.ProviderAuthData); valid {
+		// provider authData matches app's authRecordKeys,
+		// then it starts auto-link procedure.
+		//
+		// for example, if oauthAuthInfo.ProviderAuthData is {"email", "john.doe@example.com"},
+		// it will be a valid authData if authRecordKeys is [["username"], ["email"]] or [["email"]]
+		// so, the oauthAuthInfo.ProviderAuthDat can be used as a password principal authData
+		return h.authLinkUser(oauthAuthInfo)
+	}
+
+	return nil, nil
+}
+
 func (h respHandler) authLinkUser(oauthAuthInfo sso.AuthInfo) (*oauth.Principal, error) {
```

Comment:
Should `authLinkUser` return error if `h.PasswordAuthProvider.GetPrincipalByAuthData` is not `ErrUserNotFound`?